File: verify.c
/* Exact ternary-kernel check for a ±1 sign matrix (independent of verify.js).
 *
 *   cc -O2 -o verify verify.c && ./verify matrices/beta_Q29_le_15.txt
 *
 * Counts the nonzero z in {-1,0,1}^n with W z = 0 by meet-in-the-middle:
 * enumerate all 3^nL left half-assignments, sort their packed row-sum
 * signatures, then stream the 3^nR right half-assignments and binary-search
 * for exact cancellation. Zero kernel vectors  =>  the rows of W certify
 * beta(Q_n) <= q (see README.md).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

static int q, n, W[16][32];

typedef struct { uint64_t a; uint32_t b; } Sig; /* rows 0-11 in a (5 bits each), rows 12-15 in b */
typedef struct { Sig s; uint32_t idx; } Ent;

static int cmp(const void *x, const void *y) {
  const Ent *p = x, *r = y;
  if (p->s.a != r->s.a) return p->s.a < r->s.a ? -1 : 1;
  if (p->s.b != r->s.b) return p->s.b < r->s.b ? -1 : 1;
  return 0;
}

/* pack row sums (each in [-15,15], offset +15 -> [0,30] < 32) */
static Sig pack(const int *sums) {
  Sig s = {0, 0};
  for (int j = 0; j < q && j < 12; j++) s.a |= (uint64_t)(sums[j] + 15) << (5 * j);
  for (int j = 12; j < q; j++) s.b |= (uint32_t)(sums[j] + 15) << (5 * (j - 12));
  return s;
}

int main(int argc, char **argv) {
  if (argc != 2) { fprintf(stderr, "usage: %s <matrix.txt>\n", argv[0]); return 2; }
  FILE *f = fopen(argv[1], "r");
  if (!f) { perror("open"); return 2; }
  char line[128];
  while (fgets(line, sizeof line, f)) {
    if (line[0] == '#' || line[0] == '\n') continue;
    int len = (int)strcspn(line, "\r\n");
    if (n == 0) n = len;
    if (len != n || q >= 16 || n > 32) { fprintf(stderr, "malformed matrix\n"); return 2; }
    for (int i = 0; i < n; i++) {
      if (line[i] != '+' && line[i] != '-') { fprintf(stderr, "bad char\n"); return 2; }
      W[q][i] = line[i] == '+' ? 1 : -1;
    }
    q++;
  }
  fclose(f);
  int nL = (n + 1) / 2; if (nL > 15) nL = 15;
  int nR = n - nL;
  long sizeL = 1, sizeR = 1;
  for (int i = 0; i < nL; i++) sizeL *= 3;
  for (int i = 0; i < nR; i++) sizeR *= 3;

  Ent *tab = malloc((size_t)sizeL * sizeof(Ent));
  if (!tab) { fprintf(stderr, "oom\n"); return 2; }

  /* left enumeration, incremental odometer starting at z = (-1,...,-1) */
  int sums[16] = {0};
  signed char dg[16] = {0};
  for (int j = 0; j < q; j++) { sums[j] = 0; for (int i = 0; i < nL; i++) sums[j] -= W[j][i]; }
  memset(dg, 0, sizeof dg);
  for (long idx = 0; idx < sizeL; idx++) {
    tab[idx].s = pack(sums); tab[idx].idx = (uint32_t)idx;
    if (idx == sizeL - 1) break;
    int i = 0;
    while (dg[i] == 2) { dg[i] = 0; for (int j = 0; j < q; j++) sums[j] -= 2 * W[j][i]; i++; }
    dg[i]++; for (int j = 0; j < q; j++) sums[j] += W[j][i];
  }
  qsort(tab, (size_t)sizeL, sizeof(Ent), cmp);

  /* right enumeration: need left sums equal to -(right sums) */
  long matches = 0, shown = 0;
  for (int j = 0; j < q; j++) { sums[j] = 0; for (int i = nL; i < n; i++) sums[j] += W[j][i]; } /* = -(sum at z=-1) */
  memset(dg, 0, sizeof dg);
  for (long idx = 0; idx < sizeR; idx++) {
    Sig want = pack(sums);
    long lo = 0, hi = sizeL;
    while (lo < hi) { long m = (lo + hi) / 2; Ent e = tab[m];
      if (e.s.a < want.a || (e.s.a == want.a && e.s.b < want.b)) lo = m + 1; else hi = m; }
    for (long m = lo; m < sizeL && tab[m].s.a == want.a && tab[m].s.b == want.b; m++) {
      matches++;
      long a = tab[m].idx, b = idx;
      int nz = 0; char zs[40];
      for (int i = 0; i < nL; i++) { int v = (int)(a % 3) - 1; a /= 3; zs[i] = "-.+"[v + 1]; nz |= v; }
      for (int i = 0; i < nR; i++) { int v = (int)(b % 3) - 1; b /= 3; zs[nL + i] = "-.+"[v + 1]; nz |= v; }
      zs[n] = 0;
      if (nz && shown++ < 10) printf("  %s\n", zs);
    }
    if (idx == sizeR - 1) break;
    int i = 0;
    while (dg[i] == 2) { dg[i] = 0; for (int j = 0; j < q; j++) sums[j] += 2 * W[j][nL + i]; i++; }
    dg[i]++; for (int j = 0; j < q; j++) sums[j] -= W[j][nL + i];
  }
  free(tab);
  long nulls = matches - 1; /* the z = 0 pairing always matches */
  printf("matrix: %d x %d\n", q, n);
  printf("nonzero ternary kernel vectors: %ld\n", nulls);
  if (nulls == 0) {
    printf("VALID certificate: beta(Q_%d) <= %d  (%d-attempt strategy for a %d-question test)\n", n, q, q + 1, n);
    return 0;
  }
  printf("INVALID: matrix does not resolve the hypercube\n");
  return 1;
}

File: verify.js
// Exact ternary-kernel check for a ±1 sign matrix.
//
//   usage: bun verify.js <matrix.txt>     (node >= 18 works too)
//
// The matrix file holds q rows of n characters from {+,-} (lines starting
// with '#' are ignored). The script counts, exactly, the nonzero vectors
// z in {-1,0,1}^n with W z = 0, by meet-in-the-middle enumeration of the
// 3^(n/2) half-assignments. If the count is 0, the rows of W are a valid
// certificate that beta(Q_n) <= q (see README.md for the one-line lemma).
const file = process.argv[2];
if (!file) { console.error("usage: verify.js <matrix.txt>"); process.exit(2); }
const { readFileSync } = await import("fs");
const lines = readFileSync(file, "utf8").trim().split("\n").filter((l) => !l.startsWith("#"));
const n = lines[0].length;
const q = lines.length;
if (lines.some((l) => l.length !== n || /[^+-]/.test(l))) { console.error("malformed matrix"); process.exit(2); }
if (q > 15 || n > 32) { console.error("checker supports q <= 15, n <= 32"); process.exit(2); }
const W = lines.map((l) => [...l].map((c) => (c === "+" ? 1 : -1)));

// Split columns into left (nL) and right (n - nL); pack each half-assignment's
// q row-sums (offset by nL so digits stay in [0, 2*nL] < 32) into one number
// per 10 rows: 10 base-32 digits fit exactly in a float64 mantissa.
const nL = Math.min(15, Math.ceil(n / 2)), nR = n - nL;
const fq = Math.min(q, 10), aq = q - fq;
const OFF = 15; // digit offset; row sums over <=15 columns fit [-15,15]
const p32 = [...Array(fq)].map((_, j) => Math.pow(32, j));
const a32 = [...Array(aq)].map((_, r) => Math.pow(32, r));
const dKey = new Float64Array(n), dAux = new Float64Array(n);
for (let i = 0; i < n; i++) {
  let k = 0, a = 0;
  for (let j = 0; j < fq; j++) k += W[j][i] * p32[j];
  for (let r = 0; r < aq; r++) a += W[fq + r][i] * a32[r];
  dKey[i] = k; dAux[i] = a;
}
const BITS = n >= 27 ? 25 : 22, SIZE = 1 << BITS, MASK = SIZE - 1;
const tKey = new Float64Array(SIZE).fill(-1), tAux = new Float64Array(SIZE), tIdx = new Uint32Array(SIZE);
const hash = (key) => {
  const lo = key % 33554432, hi = Math.floor(key / 33554432);
  return (Math.imul(lo, 0x9e3779b1) ^ Math.imul(hi, 0x85ebca77)) & MASK;
};
const sizeL = Math.pow(3, nL), sizeR = Math.pow(3, nR);
// left half: insert every z_L with key = packed(OFF + sum_j W_L z_L)
{
  let key = 0, aux = 0; // start at z = (-1,...,-1)
  for (let j = 0; j < fq; j++) { let s = OFF; for (let i = 0; i < nL; i++) s -= W[j][i]; key += s * p32[j]; }
  for (let r = 0; r < aq; r++) { let s = OFF; for (let i = 0; i < nL; i++) s -= W[fq + r][i]; aux += s * a32[r]; }
  const dg = new Int8Array(nL);
  for (let idx = 0; idx < sizeL; idx++) {
    let h = hash(key * 31 + aux);
    while (tKey[h] !== -1) h = (h + 1) & MASK;
    tKey[h] = key; tAux[h] = aux; tIdx[h] = idx;
    if (idx === sizeL - 1) break;
    let i = 0;
    while (dg[i] === 2) { dg[i] = 0; key -= 2 * dKey[i]; aux -= 2 * dAux[i]; i++; }
    dg[i]++; key += dKey[i]; aux += dAux[i];
  }
}
// right half: for every z_R, look up z_L with W_L z_L = -W_R z_R
const nulls = [];
{
  let key = 0, aux = 0;
  for (let j = 0; j < fq; j++) { let s = OFF; for (let i = nL; i < n; i++) s += W[j][i]; key += s * p32[j]; }
  for (let r = 0; r < aq; r++) { let s = OFF; for (let i = nL; i < n; i++) s += W[fq + r][i]; aux += s * a32[r]; }
  const dg = new Int8Array(nR || 1);
  for (let idx = 0; idx < sizeR; idx++) {
    let h = hash(key * 31 + aux);
    while (tKey[h] !== -1) {
      if (tKey[h] === key && tAux[h] === aux) {
        const z = new Int8Array(n);
        let a = tIdx[h];
        for (let i = 0; i < nL; i++) { z[i] = (a % 3) - 1; a = Math.floor(a / 3); }
        let b = idx;
        for (let i = 0; i < nR; i++) { z[nL + i] = (b % 3) - 1; b = Math.floor(b / 3); }
        if (z.some((v) => v !== 0)) nulls.push(z);
      }
      h = (h + 1) & MASK;
    }
    if (idx === sizeR - 1) break;
    let i = 0;
    while (dg[i] === 2) { dg[i] = 0; key += 2 * dKey[nL + i]; aux += 2 * dAux[nL + i]; i++; }
    dg[i]++; key -= dKey[nL + i]; aux -= dAux[nL + i];
  }
}
console.log(`matrix: ${q} x ${n}`);
console.log(`nonzero ternary kernel vectors: ${nulls.length}`);
for (const z of nulls.slice(0, 10)) console.log("  " + [...z].map((v) => "-.+"[v + 1]).join(""));
if (nulls.length === 0) {
  console.log(`VALID certificate: beta(Q_${n}) <= ${q}  (${q + 1}-attempt strategy for a ${n}-question test)`);
} else {
  console.log("INVALID: matrix does not resolve the hypercube");
  process.exit(1);
}
